src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,743 @@
+  fmap (HM.mapMaybe id) $ flip HM.traverseWithKey latestCutHeaders $ \cid latestCutHeader -> do
+    let logger' = addChainIdLabel cid logger
+    bdb <- getWebBlockHeaderDb wbhdb cid
+    seekAncestor bdb latestCutHeader (fromIntegral blockHeight) >>= \case
+      -- Block exists on that chain
+      Just h -> do
+        -- Sanity check, should absolutely never happen
+        when (_blockHeight h /= blockHeight) $ do
+          exitLog logger' "expected seekAncestor behaviour is broken"
+
+        -- Confirm that PactDB is not behind RocksDB (it can be ahead though)
+        let SQLiteEnv db _ = pactConns ^?! ix cid
+        latestPactHeight <- getLatestBlockHeight db
+        when (latestPactHeight < blockHeight) $ do
+          exitLog logger' "Pact State is behind RocksDB. This should never happen."
+
+        pure (Just h)
+
+      -- Block does not exist on that chain
+      Nothing -> do
+        logFunctionText logger' Debug $ "Block " <> sshow blockHeight <> " is not accessible on this chain."
+        pure Nothing
+
+resolveLatestCutHeaders :: (Logger logger)
+  => logger
+  -> ChainwebVersion
+  -> HashMap ChainId SQLiteEnv
+  -> RocksDb
+  -> IO (BlockHeight, HashMap ChainId BlockHeader)
+resolveLatestCutHeaders logger v pactConns rocksDb = do
+  wbhdb <- initWebBlockHeaderDb rocksDb v
+  let cutHashes = cutHashesTable rocksDb
+  latestCutHeaders <- readHighestCutHeaders v (\_ _ -> pure ()) wbhdb cutHashes
+  let latestCommonBlockHeight = minimum $ fmap _blockHeight latestCutHeaders
+  headers <- limitCut logger wbhdb latestCutHeaders pactConns latestCommonBlockHeight
+  pure (latestCommonBlockHeight, headers)
+
+resolveCutHeadersAtHeights :: (Logger logger)
+  => logger
+  -> ChainwebVersion
+  -> HashMap ChainId SQLiteEnv
+  -> RocksDb
+  -> [BlockHeight] -- ^ targets
+  -> IO [(BlockHeight, HashMap ChainId BlockHeader)]
+resolveCutHeadersAtHeights logger v pactConns rocksDb targets = do
+  wbhdb <- initWebBlockHeaderDb rocksDb v
+  let cutHashes = cutHashesTable rocksDb
+  latestCutHeaders <- readHighestCutHeaders v (\_ _ -> pure ()) wbhdb cutHashes
+  forM targets $ \target -> do
+    fmap (target, ) $ limitCut logger wbhdb latestCutHeaders pactConns target
+
+-- | Compute the GrandHashes at the specified targets.
+--
+--   The targets can either be the latest BlockHeight only,
+--   or a set of specified targets (for computing many grand hashes per chain at
+--   once).
+--
+--   The list this returns is in descending order by the BlockHeight.
+--   This is to facilitate easier searching in the style of 'findGrandHash'.
+computeGrandHashesAt :: (Logger logger)
+  => logger
+  -> HashMap ChainId SQLiteEnv
+     -- ^ pact connections
+  -> [(a, HashMap ChainId BlockHeader)]
+     -- ^ Resolved targets, i.e, blockheights that are accessible per each
+     --   chain.
+     --
+     --   The 'a' is polymorphic to show that it is unused by this function,
+     --   but we keep them paired up.
```

Comment:
Why?

Reply:
We call this later with the block heights as the `a`. This function probably doesn't need to do this itself, maybe we want the caller to do the `pooledFor`, and have this function just be:
```hs
computeGrandHashesAt :: (Logger logger)
  => HashMap ChainId SQLiteEnv
  -> HashMap ChainId BlockHeader
  -> IO (HashMap ChainId Snapshot)
```